### PR TITLE
Support out-of-cluster installation

### PR DESCRIPTION
**Pull request by salasberryfin**
**What this PR does / why we need it**:

This is intended to allow a Rancher operator to choose between deploying rancher-turtles in the same or a different cluster to Rancher Manager. The current implementation of rancher-turtles defaults to an in-cluster configuration with both Rancher Manager and rancher-turtles supposed to be installed together in the same cluster.

The initial idea is to create the client based on whether a flag with a path to a kubeconfig file is passed. When this parameter is used, rancher-turtles will load the REST Config from this file and create a client for an out-of-cluster installation. If it is not, it defaults to the current in-cluster behavior.

Issue #19 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -97,6 +101,9 @@ func initFlags(fs *pflag.FlagSet) {
 
 	fs.IntVar(&concurrencyNumber, "concurrency", 1,
 		"Number of resources to process simultaneously")
+
+	fs.StringVar(&rancherKubeconfig, "rancher-kubeconfig", "",
+		"Path to a kubeconfig file. Only required if out-of-cluster.")
```
Comment (richardcase) [inline main.go:106]:
```suggestion
		"Path to the Rancher kubeconfig file. Only required if running out-of-cluster.")
```

File: main.go [hunk 2]
```diff
@@ -154,12 +161,48 @@ func setupChecks(mgr ctrl.Manager) {
...
+			os.Exit(1)
+		}
+	} else {
+		setupLog.Info("in-cluster installation of rancher-turtles")
+		c = mgr.GetClient()
+	}
+
 	if err := (&controllers.CAPIImportReconciler{
-		Client:           mgr.GetClient(),
+		Client:           c,
```
Comment (richardcase) [inline main.go:190]: We actually need 2 clients for `CAPIImportReconciler`:

- 1 for the controller itself for the CAPI Clusters
- 1 for Rancher

So maybe something like this:

```go
type CAPIImportReconciler struct {
	Client client.Client
        RancherClient client.Client
```

In the scenario where Rancher Manager and CAPI are in the same cluster then:
- **Client** is set to `mgr.GetClient()`
- **RancherClient**  is set to `mgr.GetClient()`

In the scenario for this change (i.e. where the CAPI management cluster is different to the Rancher Manager cluster) , then:
- **Client** is set to `mgr.GetClient()`
- **RancherClient** is set to client created from the kubeconfig specified via **rancherKubeconfig**

How does that sound?

Comment (salasberryfin) [reply main.go]: @richardcase updated the PR with your suggestions.

File: main.go [hunk 3]
```diff
@@ -97,6 +101,9 @@ func initFlags(fs *pflag.FlagSet) {
 
 	fs.IntVar(&concurrencyNumber, "concurrency", 1,
 		"Number of resources to process simultaneously")
+
+	fs.StringVar(&rancherKubeconfig, "rancher-kubeconfig", "",
```
[the remark this answers is not in the document]
Comment (salasberryfin) [reply main.go:105]: Yes, this has to available in the pod.

File: internal/controllers/import_controller.go [hunk 4]
```diff
@@ -257,7 +258,7 @@ func (r *CAPIImportReconciler) reconcileNormal(ctx context.Context, capiCluster
 		Namespace: capiCluster.Namespace,
 	}
 
-	remoteClient, err := r.remoteClientGetter(ctx, capiCluster.Name, r.Client, clusterKey)
+	remoteClient, err := r.remoteClientGetter(ctx, capiCluster.Name, r.RancherClient, clusterKey)
```
Comment (Danil-Grigorev) [inline internal/controllers/import_controller.go:261]: If the providers are installed outside of rancher management cluster, then likely the secret to connect to the CAPI cluster will live inside this “other then rancher manager” cluster. So here for me it makes sense to use regular in-cluster client to connect to the provisioned cluster by CAPI from the cluster where CAPI is installed.

File: main.go [hunk 5]
```diff
@@ -155,11 +162,54 @@ func setupChecks(mgr ctrl.Manager) {
...
+		RancherClient:    rancherClient,
 		WatchFilterValue: watchFilterValue,
 	}).SetupWithManager(ctx, mgr, controller.Options{MaxConcurrentReconciles: concurrencyNumber}); err != nil {
 		setupLog.Error(err, "unable to create capi controller")
 		os.Exit(1)
 	}
 }
+
+// setupRancherClient can either create a client for an in-cluster installation (rancher and rancher-turtles in the same cluster)
+// or create a client for an out-of-cluster installation (rancher and rancher-turtles in different clusters).
```
Comment (furkatgofurov7) [inline main.go:182]:
```suggestion
// setupRancherClient can either create a client for an in-cluster installation (rancher and rancher-turtles in the same cluster)
// or create a client for an out-of-cluster installation (rancher and rancher-turtles in different clusters) based on the existence of Rancher kubeconfig file.
```

**Review by furkatgofurov7 — COMMENTED** [review]
While this is clear to the internal team, it would be nice to have a small document describing what is (out-of/in)-cluster support, the difference between them and why it is needed. 

Please disregard if we have a separate issue to document it in rancher-turtles-docs

**Comment by Danil-Grigorev** [comment]
Same thought as @furkatgofurov7, but additionally we need to have some clear process to pass the kubeconfig to the pod, which could be documented in rancher-turtles-docs + not necessary in the same PR but would prefer to have e2e test coverage here to be sure.

**Comment by richardcase** [comment]
@furkatgofurov7 @Danil-Grigorev @salasberryfin  - i created #94 for us to create an ADR for this.

**Comment by furkatgofurov7** [comment]
> @furkatgofurov7 @Danil-Grigorev @salasberryfin - i created #94 for us to create an ADR for this.

Except [this](https://github.com/rancher-sandbox/rancher-turtles/pull/82#discussion_r1311217975), I have nothing major to not merge it

**Comment by salasberryfin** [comment]
Thanks for the reviews @richardcase @Danil-Grigorev @furkatgofurov7. I'll be taking on the ADR proposal to have this well documented.

**Comment by salasberryfin** [comment]
I addressed the latest comments and created a PR for the ADR #95.

**Review by richardcase — DISMISSED** [review]
This looks good to me, thanks @salasberryfin 

We will need to create a documentation issue (in racher-turtles-docs) to cover:
- Handling the RBAC, service account etc
- How to create the secret and mount it

**Comment by salasberryfin** [comment]
I created rancher-sandbox/rancher-turtles-docs#13 to track this.

**Comment by furkatgofurov7** [comment]
@salasberryfin can we rebase it?

**Comment by salasberryfin** [comment]
The changes are now up to date after rebase, please take a look.

**Review by furkatgofurov7 — APPROVED** [review]
Thanks @salasberryfin!

**Review by Danil-Grigorev — APPROVED** [review]

**Review by alexander-demicev — APPROVED** [review]

**Pull request merged**